icaruscode/Filters/FilterCRTPMTMatching_module.cc

```diff
@@ -0,0 +1,618 @@
+// Plugin Type: analyzer (Unknown Unknown)
+// File:        FilterCRTPMTMatching_module.cc
+//
+// Generated at Thu Feb 17 13:31:00 2022 by Biswaranjan Behera using cetskelgen
+// from  version .
+////////////////////////////////////////////////////////////////////////
+
+#include "art/Framework/Core/EDFilter.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "canvas/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+#include "art_root_io/TFileService.h"
+#include "canvas/Persistency/Common/FindManyP.h"
+
+// LArSoft includes
+#include "larcore/CoreUtils/ServiceUtil.h"
+#include "lardata/DetectorInfoServices/DetectorClocksService.h"
+#include "larcorealg/Geometry/GeometryCore.h"
+#include "larcore/Geometry/Geometry.h"
+
+// Data product includes
+#include "lardataobj/RecoBase/OpHit.h"
+#include "lardataobj/RecoBase/OpFlash.h"
+#include "sbnobj/Common/CRT/CRTHit.hh"
+#include "icaruscode/CRT/CRTUtils/CRTCommonUtils.h"
+#include "sbnobj/Common/Trigger/ExtraTriggerInfo.h"
+
+// C++ includes
+#include <vector>
+#include <map>
+#include <numeric>
+
+// ROOT includes
+#include "TTree.h"
+#include "TVector3.h"
+
+namespace icarus
+{
+    namespace crt
+    {
+        class FilterCRTPMTMatching;
+        struct CRTPMT
+        {
+            double tof;
+            double distance;
+            art::Ptr<sbn::crt::CRTHit> CRTHit;
+        };
+        struct MatchedCRT
+        {
+            // vector of pairs where first is the matched Time of Flight and second is the matched CRTHit
+            std::vector<CRTPMT> entering;
+            std::vector<CRTPMT> exiting;
+            int matchType;
+        };
+        struct TrajPoint
+        {
+            double X;
+            double Y;
+            double Z;
+            double T;
+        };
+    }
+}
+
+using namespace icarus::crt;
+
+bool flashInTime(double const &flashTime, int gateType, double gateDiff, double flashTimeCut)
+{
+    std::map<unsigned int, double> gateLength{
+        {1, 2.2},  // BNB
+        {2, 10.1}, // NuMI
+        {3, 2.2},  // BNB Offbeam
+        {4, 10.1}, // NuMI Offbeam
+    };
+
+    /*double vetoOffset = 4;*/
+    double activeGate = gateLength[gateType] /*- vetoOffset*/;
+
+    double relFlashTime = flashTime + gateDiff / 1000. /*- vetoOffset*/;
+    mf::LogInfo("FilterCRTPMTMatching::flashInTime") << "Gate Diff " << gateDiff / 1000 << " Ftime+gateDiff " << flashTime + gateDiff / 1000. << " " << activeGate;
+
+    return ((relFlashTime > flashTimeCut) && (relFlashTime < activeGate));
+}
+
+icarus::crt::MatchedCRT CRTHitmatched(const double &flashTime, const double flashpos[3], std::vector<art::Ptr<sbn::crt::CRTHit>> &crtHits, const double &interval)
```

Comment:
Is it actually slower to pass an int or double (ie single number) by value vs reference? Somehow I'd think it is a single memory location either way and the instructions to recover the address vs value. But I don't know the answer hence the question...

Reply:
By the book it is slower: by value is (1) take the value at the known address vs. by reference (1) take the address value at the known address and (2) take the value at that address. In practice, compiler optimisation is more effective than anything poor physicists may hope to do. Not to mention that it depends on the processor too.
The most relevant difference here is that you are asking for `interval` to have an address, so that it may be used by reference, while I may instead specify as an interval something (an "rvalue") that has no address, like a literal number `0.1`.
Except that even in this case C++ will mock up a variable for that value.